pkg/cloudprovider/aws/aws.go

```diff
@@ -133,89 +142,165 @@ func (p *provider) TerminateInstance(providerID string) error {
+			if err != nil {
+				a.logger.Info("[ReadyInstances] skip instance which failed instanceID to providerID conversion: %v", *i.InstanceId)
+				continue
+			}
+			instances[providerID] = &instance{
+				instance:      i,
+				outOfDate:     a.instanceOutOfDate(i),
+				nodeGroupName: *group.AutoScalingGroupName,
+			}
+		}
 	}
 	return instances
 }
 
-// ReadyInstances returns a slice of instances that are InService
-func (a *autoscalingGroup) ReadyInstances() (instances []cloudprovider.Instance) {
-	for _, i := range a.group.Instances {
-		if aws.StringValue(i.LifecycleState) != "InService" {
-			continue
+// NotReadyInstances returns a map of instances that are not InService
+// with providerID as key and cloudprovider.Instance as value
+func (a *autoscalingGroups) NotReadyInstances() map[string]cloudprovider.Instance {
+	instances := make(map[string]cloudprovider.Instance)
+	for _, group := range a.groups {
+		for _, i := range group.Instances {
+			if aws.StringValue(i.LifecycleState) != "InService" {
+				providerID, err := instanceIDToProviderID(*i.InstanceId, *i.AvailabilityZone)
+				if err != nil {
+					a.logger.Info("[NotReadyInstances] skip instance which failed instanceID to providerID conversion: %v", *i.InstanceId)
+					continue
+				}
+				instances[providerID] = &instance{
+					instance:      i,
+					outOfDate:     a.instanceOutOfDate(i),
+					nodeGroupName: *group.AutoScalingGroupName,
+				}
+			}
 		}
-		instances = append(instances, &instance{
-			instance:  i,
-			outOfDate: a.instanceOutOfDate(i),
-		})
 	}
 	return instances
 }
 
-// NotReadyInstances returns a slice of instances that are not InService
-func (a *autoscalingGroup) NotReadyInstances() (instances []cloudprovider.Instance) {
-	for _, i := range a.group.Instances {
-		if aws.StringValue(i.LifecycleState) != "InService" {
-			instances = append(instances, &instance{
-				instance:  i,
-				outOfDate: a.instanceOutOfDate(i),
-			})
+// getInstanceNodeGroupByInstanceID finds the right autoscaling group for an instance based on
+// instance ID matches one instance ID from one of the node groups inside nodeGroupsList
+func (a *autoscalingGroups) getInstanceNodeGroupByInstanceID(instanceID string) (*autoscaling.Group, error) {
+	var targetNodeGroup *autoscaling.Group
+	foundNodeGroup := false
+	for _, group := range a.groups {
+		for _, i := range group.Instances {
+			if *i.InstanceId == instanceID {
+				targetNodeGroup = group
+				foundNodeGroup = true
+				break
+			}
+		}
+		if foundNodeGroup {
+			break
 		}
 	}
-	return instances
+	if foundNodeGroup {
+		return targetNodeGroup, nil
+	}
+	return nil, fmt.Errorf("failed to found target node group for instance with ID: %v", instanceID)
+}
+
+// getInstanceNodeGroupByGroupName finds the right autoscaling group for an instance based on
+// nodeGroup name passed in
+func (a *autoscalingGroups) getInstanceNodeGroupByGroupName(nodeGroup string) (*autoscaling.Group, error) {
+	if nodeGroup != "" {
```

Comment:
`nit` nesting and happy path leftmost

```
if nodeGroup != "" {
   return nil, fmt.Errorf("failed to found target node group for instance with group: %v", nodeGroup)
}

for ...
```

Reply:
Actually that last error serves two different conditions: a) `nodeGroup == ""` b) no matching group found inside the existing groups list. 
But I can split that and add another error for the first condition.